drivers/regulator/regulator_adp5360.c

```diff
@@ -0,0 +1,338 @@
+	uint8_t val;
+	int ret;
+
+	if (!config->desc->has_modes) {
+		return -ENOTSUP;
+	}
+
+	ret = i2c_reg_read_byte_dt(&pconfig->i2c, config->desc->cfg_reg, &val);
+	if (ret < 0) {
+		return ret;
+	}
+
+	*mode = (val & ADP5360_BUCK_CFG_BUCK_MODE_MSK) >> ADP5360_BUCK_CFG_BUCK_MODE_POS;
+
+	return 0;
+}
+
+static int regulator_adp5360_enable(const struct device *dev)
+{
+	const struct regulator_adp5360_config *config = dev->config;
+	const struct regulator_adp5360_pconfig *pconfig = config->p->config;
+
+	return i2c_reg_update_byte_dt(&pconfig->i2c, config->desc->cfg_reg, ADP5360_BUCK_CFG_EN_MSK,
+				      1U);
+}
+
+static int regulator_adp5360_disable(const struct device *dev)
+{
+	const struct regulator_adp5360_config *config = dev->config;
+	const struct regulator_adp5360_pconfig *pconfig = config->p->config;
+
+	return i2c_reg_update_byte_dt(&pconfig->i2c, config->desc->cfg_reg, ADP5360_BUCK_CFG_EN_MSK,
+				      0U);
+}
+
+static int regulator_adp5360_init(const struct device *dev)
+{
+	const struct regulator_adp5360_config *config = dev->config;
+	const struct regulator_adp5360_pconfig *pconfig = config->p->config;
+	int ret;
+	uint8_t val, nval, msk;
+
+	regulator_common_data_init(dev);
+
+	if (!device_is_ready(config->p)) {
+		return -ENODEV;
+	}
+
+	/* apply optional delay */
+	msk = 0U;
+	nval = 0U;
+
+	ret = i2c_reg_read_byte_dt(&pconfig->i2c, config->desc->out_reg, &val);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if (config->dly_idx >= 0) {
+		msk |= ADP5360_BUCK_OUTPUT_DLY_MSK;
+		nval |= ((uint8_t)config->dly_idx << ADP5360_BUCK_OUTPUT_DLY_POS) &
+			ADP5360_BUCK_OUTPUT_DLY_MSK;
+	}
+
+	if (msk != 0U) {
+		ret = i2c_reg_write_byte_dt(&pconfig->i2c, config->desc->out_reg,
+					    (val & ~msk) | nval);
+		if (ret < 0) {
+			return ret;
+		}
+	}
+
+	/* apply optional initial configuration */
+	msk = 0U;
+	nval = 0U;
+
+	ret = i2c_reg_read_byte_dt(&pconfig->i2c, config->desc->cfg_reg, &val);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if (config->ss_idx >= 0) {
+		msk |= ADP5360_BUCK_CFG_SS_MSK;
+		nval |= ((uint8_t)config->ss_idx << ADP5360_BUCK_CFG_SS_POS) &
+			ADP5360_BUCK_CFG_SS_MSK;
+	}
+
+	if (config->ilim_idx >= 0) {
+		if (config->desc->has_modes) {
+			msk |= ADP5360_BUCK_CFG_BUCK_ILIM_MSK;
+			nval |= ((uint8_t)config->ilim_idx << ADP5360_BUCK_CFG_BUCK_ILIM_POS) &
+				ADP5360_BUCK_CFG_BUCK_ILIM_MSK;
+		} else {
+			msk |= ADP5360_BUCK_CFG_BST_ILIM_MSK;
+			nval |= ((uint8_t)config->ilim_idx << ADP5360_BUCK_CFG_BST_ILIM_POS) &
+				ADP5360_BUCK_CFG_BST_ILIM_MSK;
+		}
+	}
+
+	if (config->stp_en) {
+		msk |= ADP5360_BUCK_CFG_STP;
+		nval |= ADP5360_BUCK_CFG_STP;
+	}
+
+	if (config->dis_en) {
+		msk |= ADP5360_BUCK_CFG_DISCHG;
+		nval |= ADP5360_BUCK_CFG_DISCHG;
+	}
+
+	if (msk != 0U) {
+		ret = i2c_reg_write_byte_dt(&pconfig->i2c, config->desc->cfg_reg,
+					    (val & ~msk) | nval);
+		if (ret < 0) {
+			return ret;
+		}
+	}
+
+	return regulator_common_init(dev, !!(val & ADP5360_BUCK_CFG_EN_MSK));
```

Comment:
Do you really need the "!!"? The compiler should do it for you anyway, I guess this makes it explicit but it does not seem like it's done much in the code base.

Reply:
well, that's true, but then we need a cast to make it explicitely boolean, no?

Reply:
Do you get a compiler warning?

Reply:
In theory this violates rule 10.3 https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html, but usage of `!` with unsigned as well with 10.1... And... casting to `bool` violates 10.5. Lots of fun.

Reply:
changed to

```c
	return regulator_common_init(dev, (val & ADP5360_BUCK_CFG_EN_MSK) != 0U);
```

that seems to pass misra compliance

Reply:
🙄